tests/5.0/task/test_task_detach.c

```diff
@@ -18,42 +22,47 @@
 
 #define N 1024
 
+int test_callback(omp_event_handle_t event) {
+  omp_fulfill_event(event);
+  return 1;
+}
+
 int test_task_detach() {
   OMPVV_INFOMSG("test_task_detach");
   int errors = 0, x = 0, y = 0;
   int num_threads = -1, record_x = -1, record_y = -1;
-  omp_event_handle_t *flag_event;
+  omp_event_handle_t flag_event;
 
-#pragma omp parallel num_threads(2) default(shared)
+#pragma omp parallel
+#pragma omp single
   {
-    if (omp_get_thread_num() == 1 || omp_get_num_threads() < 2) {
-      num_threads = omp_get_num_threads();
-      x++;
-      omp_fulfill_event(flag_event);
+#pragma omp task depend(out: y) detach(flag_event)
+    {
+      y++;
     }
-    if (omp_get_thread_num() == 0) {
-#pragma omp task detach(flag_event)
-      {
-        y++;
-      }
-#pragma omp taskwait
+#pragma omp task
+    {
+      x = test_callback(flag_event);
+    }
+#pragma omp task depend(inout: y)
+    {
       record_x = x;
```

Comment:
Isn't there a potential race condition for `x`? `y` gets updated and this `record_* =` task waits for it as does `test_callback`, but the `x = ` assignment could happen in parallel with `record_x = ...`. Or do I miss something?

Reply:
Hmm I see the concern. I guess those changes would bring us back to the original problem of making the test not really use the detach feature.

Reply:
Although my understanding is that the first task, because of its detach clause, will not be considered complete for the purposes of depend until both its associated block is complete and the event is fulfilled. Wouldn't that mean the third task will not start execution until the callback is executed and y is updated? Although I could see an issue with the third task starting before the write to x in the second task is completed.

Here's the 5.0 spec on the detach clause (p. 137, third paragraph):

> If a detach clause is present on a task construct a new event allow-completion-event is created.
> The allow-completion-event is connected to the completion of the associated task region. The
> original event-handle will be updated to represent the allow-completion-event event before the task
> data environment is created. The event-handle will be considered as if it was specified on a
> firstprivate clause. The use of a variable in a detach clause expression of a task
> construct causes an implicit reference to the variable in all enclosing constructs.

What if we change the body of the second task to the the following:

x = 1;
#pragma omp flush
test_callback(flag_event);

This way x is updated before the event is fulfilled. We'd also need to add a flush before record_x in the third task.

Reply:
I made this update and it passes Clang, should appear on this PR now. @tob2 do you think this is correct/race-free?

Reply:
Tested with GCC by a colleague; it works with -O0 but not with -O3.

My built does not support it, but when I look at the dump (-fdump-tree-optimized -O0) of a slightly massaged code, I think that is because of a race condition with
```C
#pragma omp task
    {   x = 1; ... }
#pragma omp task depend(inout: y)
    { ...; record_x = x; ... }
```
which GCC turns into something like the following (pseudo code)
```C
args2.x = &x;  // pass by reference - as the value is modified: 'x = 1'
__builtin_GOMP_task (test_task_detach._omp_fn.2, &args2);
args3.x = x; // pass by value - as the value is only read: 'record_x = x'
__builtin_GOMP_task (test_task_detach._omp_fn.3, &args3);
```
And when the task-calling function returns before the task thread has set `x`, it fails – as doing the assignment is quick and due to pass-by-value, the 'omp flush' does not help.

I think GCC's optimization is fine, but I did not (re)check the spec – if so, the race condition is already in the testcase and not only caused by the compiler.

Reply:
Re-reading this patch, I think the code does *not*  have a race condition and your patch is *fine*.

Thus: *LGTM*